downstairs/src/region.rs

```diff
@@ -1057,17 +1163,33 @@ impl Extent {
         }
         tx.commit()?;
 
+        // Buffer writes for fewer syscalls. The 65536 buffer size here is
+        // chosen somewhat arbitrarily. We could reimplement this ourselves
+        // without BufWriter but I don't see a reason to at the moment.
+        let mut buffered_writer = BufWriter::with_capacity(65536, &inner.file);
+
+        let mut next_block_in_run = u64::MAX;
         for write in writes {
-            if writes_to_skip.contains(&write.offset.value) {
-                assert!(only_write_unwritten);
+            let block = write.offset.value;
+            if writes_to_skip.contains(&block) {
+                debug_assert!(only_write_unwritten);
                 continue;
             }
-            let byte_offset = write.offset.value * self.block_size;
 
-            inner.file.seek(SeekFrom::Start(byte_offset))?;
-            inner.file.write_all(&write.data)?;
+            // Seeking flushes the buffer so only do it when we need to
+            if block != next_block_in_run {
+                buffered_writer
+                    .seek(SeekFrom::Start(block * self.block_size))?;
+            }
+
+            buffered_writer.write_all(&write.data)?;
+
+            next_block_in_run = block + 1;
         }
 
+        // writes remaining buffer without calling flush() on the file handle.
+        drop(buffered_writer);
```

Comment:
The following comment from [BufWriter's source](https://doc.rust-lang.org/src/std/io/buffered/bufwriter.rs.html#23-27) makes me very anxious:

```
/// It is critical to call [`flush`] before `BufWriter<W>` is dropped. Though
/// dropping will attempt to flush the contents of the buffer, any errors
/// that happen in the process of dropping will be ignored. Calling [`flush`]
/// ensures that the buffer is empty and thus dropping will not even attempt
/// file operations.
```

Based on that comment, I would even say that this means we can't use `BufWriter` here. Previously, the downstairs would return errors from each individual `inner.file.write_all`, where here they are being ignored. We have to know when there are errors and propagate them up to the Upstairs (and eventually Nexus).

Reply:
yeah that's fair, i'll rewrite with my own buffering logic